### ISSUE TITLE
Dynamically change resource controls for user-.slice

### ISSUE THREAD

**Issue by benzea** [systemd/systemd#15028]
This came up in discussions about setting resource guarantees for userspace processes. I'll update this ticket with more information later.

But in general, I believe we will want to reserve resources for `user-.slice` so that the users session can allocate them appropriately. However, we also need to ensure that an inactive session does not get such an allocation (inactive user e.g. the GDM greeter).

I suspect this logic should live in `systemd-logind`. Though I am not sure how the interaction with the usual resource control parameters should look like.

Relevant e-mails:
 * https://lore.kernel.org/linux-mm/4d3e00457bba40b25f3ac4fd376ba7306ffc4e68.camel@sipsolutions.net/ (initial mail floating the idea)
 * https://lore.kernel.org/linux-mm/20200305145554.GA5897@mtj.thefacebook.com/ (reply, should have most of the relevant information)
 * https://lore.kernel.org/linux-mm/CAOS58YM-HtmxwD7Q7g6ZzGsOJ_vWjHwb=7qpUwuZQEdeRrBifw@mail.gmail.com/ (further important notes from Tejun)

**Comment by Werkov** [comment]
Hi @benzea, thanks for bringing this up.

I'd start with the steps:

1. Figure out where (*L*) the processes needing protection live
2. Figure out what are their needs
    - configure these values to the closest (leaf) groups (assuming some reduction factor as mentioned by Tejun)
3. Distribute the sufficient amount from root to *L*

In a simple case, this could be done by a static configuration. Making this dynamic can affect each of these steps (new/released user session; shared services whose demands grow/shrink; prioritizing when overcommit happens) and the question is what are the events that cause the changes and what should be the response.

Personally, I think that the static configs shouldn't be carried by upstream but be a distro decision. For the dynamic case, systemd-logind looks like a viable starting point for the changes notifications [1], however, I think the resource implications should be done elsewhere to make it configurable (e.g. how much should a new SSH session of a user with backgrounded GUI account for in comparison with a foregrounded user GUI).

What are the use cases that you consider? (In relation to the point 1 above and the relevant change events.)

[1] Note that pam_systemd already accepts some (per-session) resource arguments, also `user-$UID.slice` itself can have some resource attributes controlled via [user records](https://systemd.io/USER_RECORD/) (although top-down distributed resources would need an allowance from ultimate top).

**Comment by benzea** [comment]
Hi @Werkov,

 1. We want to protect `session.slice` (see !14846) which contains the main components of the user's graphical session. Other scenarios may be plausible, but are not relevant from my point of view.
 2. Yep, though I am hoping that with 5.5.7 we get the `memory_recursiveprot` mount option enabled, simplifying this task. The kernel should just do the best it can for the whole of `session.slice`. As said elsewhere, my hope is to measure KDE/GNOME using `DAMON` or `senpai` and choose reasonable limits from that.
 3. A policy that assigns this dynamically depending on the *user configurable* load L is complicated. A static assignment is much easier in my view and also completely sufficient.

So, my idea is that we take a very simple approach. i.e.:

 1. Define an (arbitrary) per-graphical session memory guarantee (`per_seat_alloc`)
 2. Assign `n_seats * per_seat_alloc` to `user.slice`.
 3. Keep track of the active session on each seat (possibly limit this to graphical/greeter sessions).
 4. Assign `per_seat_alloc` to `user-X.slice` if the user has an active session on one of the seats.

I agree that a static configuration is already pretty good. The main reason for proposing a dynamic solution is that it is reasonably common to have an inactive greeter session lingering in the background. It would be quite bad to effectively `mlock` such an inactive background session.

> Personally, I think that the static configs shouldn't be carried by upstream but be a distro decision.

Yes, these policies need to be distribution specific and will also differ between a server and desktop install. I am not sure where they should live, it could become part of `systemd-oomd` or `systemd-logind` or we could do a separate policy daemon either inside or outside of systemd.

> Note that pam_systemd already accepts some (per-session) resource arguments, also `user-$UID.slice` itself can have some resource attributes controlled via user records

I don't think the per-session stuff is relevant for us. Because I suspect that will affect the `session-X.scope`, while our graphical sessions are running outside of these scopes. I had not see the `user records` feature, thanks for the pointer. I don't see how they can help in a scenario where we want to keep the active user session responsive during low memory situations.

**Comment by benzea** [comment]
I have pushed https://gitlab.freedesktop.org/benzea/uresourced which implements something like this (tested on a GNOME session on F32).

**Comment by Werkov** [comment]
Nice job. (I'm replying here a) for context, b) since I'm not aware if there's more suitable place.)

- Enabling controllers by setting default weights
  - I find it inelegant to type out the value that should be implicit. 
  - Is that on purpose? (Such as [autogroup](https://man7.org/linux/man-pages/man7/sched.7.html).)

- `uresourced.service:Before=user@.service` doesn't work as intended (systemd v245)
  - You may use `After=uresourced.service` drop-in for `user@.service`

- `Slice=` drop-in not applicable to running units
  - You explain how to achieve runtime enabling, however, moving units between slices equals cgroup migration and that's not supported (as expected) with cgroup v2
  - only complete restart of user instance (and its units) will help, daemon-reload doesn't help with this and shouldn't be necessary when only controller attributes are modified

- `[Global]/MaxMemoryLow=`
  - Is it any better than fixed `user.slice:MemoryLow=`?
    - If the `ActiveUser.MemoryLow * active_users` product is above it, it will be enforced by kernel anyway and if it's below, why to give up possible protection (assuming the `memory_recursiveprot`)?

- What is the reasoning behind the user instance uresourced?

FWIW, since I'd like to play with it, I made an [openSUSE RPM](https://build.opensuse.org/package/binaries/home:mkoutny:uresourced/uresourced/openSUSE_Tumbleweed).

**Comment by benzea** [comment]
Uh, maybe this would have been better in a uresourced issue? :)

Anyway, I'll try to answer:

 * Not sure if there is a better way to enable the controller. IIRC you cannot set `Delegate=` on a slice unit. Though, to be honest, I didn't try to find an alternative solution.
 * Oh, I guess I need to add a drop-in for `user@.service` then. Thanks for the pointer!
 * Yes, that is why I started suggesting a reboot. (It should be sufficient to completely log out and wait for 10s, but try to explain that to people.)
 * Well, I expect that we are going to enable `memory_recursiveprot`. And I am not sure that "bypass" configurations are a great idea overall for `memory.low` (i.e. I expect file caches might end up being accounted to `user.slice` after a user session closes). See #16559.
   * That said, I could imagine adding a new `MaxMemoryLow` systemd feature. Which would effectively set `MemoryLow=min(MaxMemoryLow, sum(children.MemoryLow) + own.MemoryLow)`.
 * The user instance of `uresourced` is currently just a flag to know that the user is running a systemd enabled session. i.e. the system daemon queries whether the corresponding cgroup exists. If it does, it sets `MemoryLow` on `user@X.service` (if not, it is only set on `user-X.slice`).

**Comment by Werkov** [comment]
Yes, at this point it makes sense to move the discussion to uresourced's tracker. I posted the new issues there. Ad `MaxMemoryLow=`, I'd keep it as it's now and change it depending on practical experiences.

Does this (systemd) issue still need to be open?

**Comment by benzea** [comment]
> Does this (systemd) issue still need to be open?

I suspect not. We may want another discussion on how to better integrate with systemd eventually though.
